src/backend/distributed/commands/alter_table.c

```diff
@@ -1223,8 +1223,14 @@ CreateDistributedTableLike(TableConversionState *con)
 		newShardCount = con->shardCount;
 	}
 
+	/*
+	 * To get the correct column name, we use the original relation id, not the
+	 * new relation id. The reason is that the cached attributes of the original
+	 * and newly created tables are not the same if the original table has
+	 * dropped columns (dropped columns are still present in the attribute cache)
+	 */
 	char *distributionColumnName =
-		ColumnToColumnName(con->newRelationId, (Node *) newDistributionKey);
+		ColumnToColumnName(con->relationId, (Node *) newDistributionKey);
```

Comment:
what if `con->relationId` also has a dropped column, do we hit a similar issue?

Reply:
This issue _is_ hit if `con->relationId` has a dropped column. Did you mean to say something else? con->newRelationId doesn't have any dropped columns because it is newly created.

Reply:
Ok, reading the code a bit more, I realized that we have:
```C
	con->distributionKey =
		BuildDistributionKeyFromColumnName(con->relationId, con->distributionColumn,
										   NoLock);
```

So, what you do is probably right. Still, dropped columns are tricky, and we had so many bugs in the past. Would you mind doing some more extensive testing: 

For example, can we also change distribution column of a table whose column is dropped? Can that column be before/after the dropped column? And/or, the dropped column is dropped even before the table is distributed?

Reply:
I think we can backport this fix all the way to 10.0

Reply:
> I think we can backport this fix all the way to 10.0

I guess up to 11.0 should be fine. Otherwise, you need to handle enterprise repo as well, which I'd probably avoid unless someone explicitly asks ;)